```{r tsne data,  echo=FALSE,error=FALSE,message=FALSE,warning=FALSE}
library(plotly)
rm(gg, m, p1, p3, T)
my.col   <- readRDS("/data/MDATA/NormRcode/FixedCNSset/mycols.rds");
umap.res <- read.csv(file.path(batchdirout, "umap_2-PCA5x1.txt"), sep ="\t");
myinputdir = "/data/MDATA/compass/iScan_raw/";
columns <-c("RFpurity.ABSOLUTE", "RFpurity.ESTIMATE", "LUMP", "CNS.MCF.score", "CNS.Subclass.score")
umap.res[, columns] <- lapply(columns, function(x) as.numeric(umap.res[[x]]))
umap.res <- umap.res %>% 
    mutate(across(c(RFpurity.ABSOLUTE, RFpurity.ESTIMATE, LUMP, CNS.MCF.score, CNS.Subclass.score), ~ round(., 2)))
metarow  <- subset(umap.res, Study == "compass");

#cluster labels
dim.center = aggregate(subset(umap.res, select= c("x1", "y1", "x2", "y2")), list(umap.res$NCI_METRIC), median);
colnames(dim.center)[1] = "cancer";
xy1_annotations = lapply(unique(dim.center$cancer), function(x){
    idx = which(dim.center$cancer == x)
    ann = list(showarrow = F, x = dim.center$x1[idx], y = dim.center$y1[idx],
    text = paste0("<b>",dim.center$cancer[idx],"</b>"),
    bgcolor = alpha("white", 0.5),
    font = list(size = 10, family = "helvetica", color = my.col[dim.center$cancer[idx]]));
    ann;
})
xy2_annotations = lapply(unique(dim.center$cancer), function(x){
    idx = which(dim.center$cancer == x)
    ann = list(showarrow = F, x = dim.center$x2[idx], y = dim.center$y2[idx],
    text = paste0("<b>",dim.center$cancer[idx],"</b>"),
    bgcolor = alpha("white", 0.5),
    font = list(size = 10, color = my.col[dim.center$cancer[idx]]));
    ann;
})


```

---
title: '![`r paste(metarow$Sample)`](/data/MDATA/NormRcode/FixedCNSset/NCIlogo.png){width=45%}'
date:  '`r paste(metarow$Sample," :: ", metarow$idat_filename, " ")` '
output: html_document
fontsize: 8pt
---

```{r disclaimer, warning=F, message=F, echo=F}
DSTXT = kableExtra::kbl(paste("  Created on:", Sys.Date()), 
   col.names = gsub("[.]", " ", "Classification using methylation profiling is a research tool under development, it is not verified and has not been clinically validated. Implementation of the results in a clinical setting is in the sole responsibility of the treating physician.")) %>%
  kableExtra::kable_styling(bootstrap_options = "condensed", font_size = 8)

```

```{r umap1 code, echo=FALSE,error=FALSE,message=FALSE,warning=FALSE}
p1 = plot_ly(x = umap.res$x1, y  = umap.res$y1, 
             text = paste0(umap.res$Sample, " : ",umap.res$CNS.Subclass," : ", umap.res$CNS.Subclass.score,
                           "<br>Hist: ", umap.res$Histology), 
             color = umap.res$NCI_METRIC, colors = my.col, width = 800, height = 700,
             type="scatter", mode = "markers", marker = list(size = 6, line = list(width=0.01)),
             symbols = c('dot')) %>%
  layout(showlegend = TRUE, legend = list(font = list(size = 8)),
         annotations = xy1_annotations,
         xaxis = list(title = "umap 1", zeroline = FALSE),
         yaxis = list(title = "umap 2", zeroline = FALSE)) %>%
  add_annotations(x = umap.res[umap.res$Study == "compass", "x1"],
                  y = umap.res[umap.res$Study == "compass", "y1"],
                  text = umap.res[umap.res$Study == "compass", "Sample"],
                  showarrow = TRUE,
                  arrowcolor='black',
                  arrowhead = 6,
                  arrowsize = 1,
                  xref = "x",
                  yref = "y",
                  font = list(color = 'black',
                             family = 'arial',
                             size = 12))

```

```{r SmallTable code, echo=FALSE,error=FALSE,message=FALSE,warning=FALSE}
# MT = knitr::kable(t(metarow[, -c(1:7,16:19,22)]), col.names = gsub("[.]", " ", metarow$Sample))

MT = kableExtra::kbl(t(metarow[, -c(1:7,16:19,22)]), col.names = gsub("[.]", " ", metarow$Sample)) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped","hover", "condensed"), font_size = 8)
  
```


```{r densmap code, echo=FALSE,error=T,message=FALSE,warning=FALSE}

m<-highlight_key(umap.res[,c(
        "idat_filename",
        "x2","y2",
        "Sample",
        "Age",
        "Sex", 
        "material_prediction",
        "RFpurity.ABSOLUTE",
        "LUMP",
        "Study",
        "Location_general",  
        "Location_specific",  
        "Histology",
        "Molecular", 
        "NCI_METRIC",         
        "CNS.MCF",
        "CNS.MCF.score", 
        "CNS.Subclass",
        "CNS.Subclass.score")])

p3<-ggplot(m,aes(x2,y2)) +
#            geom_point(aes(color = CNS.Subclass)) +
            geom_point(aes(color = NCI_METRIC, label = Histology, label2 = Molecular, label3 = Sample)) +
            theme_minimal() +
            scale_color_manual("Methylation class", 
                                values = my.col[sort(names(my.col))], 
                                guide = guide_legend(override.aes = list(shape = 16))) 

gg<-highlight(ggplotly(p3),on = "plotly_selected", off = "plotly_relayout")

T<-crosstalk::bscols(list(gg  %>%
          add_annotations(x = subset(umap.res, Study == "compass")$x2,
          y = subset(umap.res, Study == "compass")$y2,
          text = subset(umap.res, Study == "compass")$Sample,
          showarrow = TRUE,
          arrowcolor='black',
          arrowhead = 6,
          arrowsize = 1,
          xref = "x",
          yref = "y",
          font = list(color = 'black', family = 'arial',size = 16)),
          
          DT::datatable(m, rownames=F, 
                   extensions = 'Buttons',
                   options = list(autoWidth = TRUE,
                      lengthMenu = c(5, 20, 50, 100),
 #		                  scrollY = '1000px',
                      scrollY = T, 
                      scrollX = T,
                      columnDefs = list(list(visible=FALSE, targets=c(0,1,2,4:9,11,13,16))),
 				              paging = TRUE,
 				              searching = TRUE,
 				              fixedColumns = TRUE,
 				              ordering = TRUE,
 				              dom = 'Blfrtip',
 				              buttons = c('colvis', 'copy', 'csv', 'excel'))) %>%

          DT::formatStyle(columns = c(1:3,6:17), fontSize = '70%')))

```


```{r dbscan code, echo=FALSE,error=FALSE,message=FALSE,warning=FALSE}
library(dbscan,  lib = "/data/turakulovr2/R_libs")
cl <- hdbscan(umap.res[,c("x1","y1")], minPts = 21)
## minPts = 21-23 is the closest number to the original number of labels
umap.res$cluster         <- cl$cluster;
umap.res$membership_prob <- cl$membership_prob
umap.res$outlier_scores  <- cl$outlier_scores

p2 = plot_ly(x = umap.res$x1, y  = umap.res$y1, 
             text = paste0(umap.res$Sample, " : ",umap.res$CNS.Subclass," : ", umap.res$CNS.Subclass.score,
                           "<br>Hist: ", umap.res$Histology, "<br>ClustNscore: ", umap.res$cluster," | ", round(umap.res$membership_prob,2)), 
             color = umap.res$cluster+1, 
			 width = 800, height = 700,
             type="scatter", mode = "markers",
			 marker = list(size  = ifelse(umap.res$cluster == 0,  5,  7), # Decrease size of noise dots
			              symbol = ifelse(umap.res$cluster == 0,  19, "circle"), # Mark noise as star
			              line   = list(width=0.01)),
             symbols = c('dot')) %>%
  layout(showlegend = TRUE, legend = list(font = list(size = 8)),
         annotations = xy1_annotations,
         xaxis = list(title = "umap 1", zeroline = FALSE),
         yaxis = list(title = "umap 2", zeroline = FALSE)) %>%
  
  add_annotations(x = umap.res[umap.res$Study == "compass", "x1"],
                        y = umap.res[umap.res$Study == "compass", "y1"],
                        text = umap.res[umap.res$Study == "compass", "Sample"],
                        showarrow = TRUE,
                        arrowcolor='black',
                        arrowhead = 6,
                        arrowsize = 1,
                        xref = "x",
                        yref = "y",
                        font = list(color = 'black',
                                     family = 'arial',
                                     size = 12))

sid = metarow$Sample;
samplecluster = umap.res[umap.res$Study=="compass", "cluster"];
if(samplecluster == 0){
	samplecluster = "Outlier";
	cl.score = "Not aplicable";
	}else{
	cl.score = cl$cluster_scores[umap.res[umap.res$Study=="compass", "cluster"]]
};

TDBS <- data.frame( sid=c(
   umap.res[umap.res$Study=="compass", "membership_prob"],
   umap.res[umap.res$Study=="compass", "outlier_scores"],
   samplecluster,
   cl.score,
   cl$hc$method, 
   cl$hc$dist.method,
   dim(cl$hc[[1]])[1], 
   cl$minPts,
   last(names(cl$cluster_scores)),
   sum(cl$cluster==0)));
row.names(TDBS) <- c("Membership probability",  
                     "Outlier scores", 
					 "Cluster assigned", 
					 "Cluster score", 
					 "Method used in DBSCAN", 
					 "Distance Method", 
					 "Number of subjects (total)", 
					 "Minimum size of clusters", 
					 "Total number of clusters by DBSCAN", 
					 "Number of outliers (black triangles)");
colnames(TDBS)<-sid;

## distances to the cluster centers are precalculated in norm script
tophits <- rownames(distances)[order(distances[ , metarow$Sample])]

TL1 <- as.vector(distances[tophits[1], umap.res$Sample[umap.res$NCI_METRIC == tophits[1]]])
TL2 <- as.vector(distances[tophits[2], umap.res$Sample[umap.res$NCI_METRIC == tophits[2]]])
TL3 <- as.vector(distances[tophits[3], umap.res$Sample[umap.res$NCI_METRIC == tophits[3]]])

Z1 <- (distances[tophits[1], metarow$Sample] - mean(TL1))/sd(TL1)
Z2 <- (distances[tophits[2], metarow$Sample] - mean(TL2))/sd(TL2)
Z3 <- (distances[tophits[3], metarow$Sample] - mean(TL3))/sd(TL3)
P1 <- pnorm(q=Z1, mean = mean(TL1), sd = sd(TL1), lower.tail=FALSE)
P2 <- pnorm(q=Z2, mean = mean(TL2), sd = sd(TL2), lower.tail=FALSE)
P3 <- pnorm(q=Z2, mean = mean(TL3), sd = sd(TL3), lower.tail=FALSE)


```

### {.tabset}



#### UMAP (unsupervised)
```{r umap figure1, echo=FALSE, error=FALSE, message=FALSE, warning=FALSE}
p1

MT

```

```{r umap CNV, echo=FALSE, error=FALSE, message=FALSE, warning=FALSE, fig.width = 19}
message("\n\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\nBuilding copy number chromosome picture...");

gfile = as.character(paste0(metarow$idat_filename, "_Grn.idat"))
cbarcode = gsub("[_R].+", "", metarow$idat_filename, perl=T);
fsize = file.info(file.path(paste0(myinputdir, cbarcode), gfile))$size
if(fsize > 10000000){
	arraytype = "EPIC";
}else{
	arraytype = "450k";
}
message(gfile, " >>> Looks like >>> ", arraytype)

# load libraries
#suppressMessages(library(conumee, lib="/data/turakulovr2/R_libs"))
#suppressMessages(library(minfi, lib="/data/turakulovr2/R_libs"))
suppressMessages(library(conumee))
suppressMessages(library(minfi))

RGset <- read.metharray(file.path(paste0(myinputdir, cbarcode), gfile),verbose=TRUE)
Mset  <- preprocessRaw(RGset)

# load conumee annoatation object according to array type
if(arraytype == "EPIC"){
  load(file.path("/data/MDATA/idat/idat4LGGnPA/CNV_data","conumee_annotation_EPIC_B6.2019-11-29.RData"))
  # load conumee reference male ( female needs different file)
  load(file.path("/data/MDATA/idat/idat4LGGnPA/CNV_data","CNanalysis6_conumee_REF_M.2018-09-19.RData"))

    cndata <- CNV.load(Mset)
    x <- CNV.fit(cndata, refM_epic, annoEPICxy)

}else{
	load(file.path("/data/MDATA/idat/idat4LGGnPA/CNV_data","CNanalysis4_conumee_ANNO.vh20150715.RData"))
	load(file.path("/data/MDATA/idat/idat4LGGnPA/CNV_data","CNanalysis4_conumee_REF-M.vh20150715.RData"))
	cndata <- CNV.load(Mset)
	x <- CNV.fit(cndata, refM.data, annoXY)
}
x <- CNV.bin(x)
x <- CNV.detail(x)
x <- CNV.segment(x)

CNV <- CNV.genomeplot(x, chrX=F, chrY=F );

```


```{r umap, warning=F, message=F, echo=F}
kableExtra::kbl(c("Classification using methylation profiling is a research tool under development, it is not verified and has not been clinically validated. Implementation of the results in a clinical setting is in the sole responsibility of the treating physician.", paste("  Created on:", Sys.Date())), col.names = gsub("[.]", " ", paste0(metarow$Sample, " has no copy number for sex chromosomes calculated."))) %>%
  kableExtra::kable_styling(bootstrap_options = "condensed", font_size = 8)

```

#### DENSMAP
```{r display_data, warning=F, message=F, echo=F}
T

DSTXT

```

#### DBSCAN
```{r dbscan, warning=F, message=F, echo=F}
p2

kableExtra::kbl(TDBS) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped","hover", "condensed"), font_size = 8)


par(cex.main=0.75)
boxplot(TL1, TL2, TL3,
	ylim = c(0, max(distances[tophits[1:3], metarow$Sample])),
	main = paste0("Distances from ", sid, " to the three nearest cluster centers."),
	names = paste(tophits[1:3], c("\nZ: ","\nZ: ","\nZ: "), round(c(Z1,Z2,Z3),4)), outline=F);
points(x=1:3, y=distances[tophits[1:3], metarow$Sample], bg="red", cex=3, pch=21);
text(x=1:3, y=distances[tophits[1:3], metarow$Sample], labels=paste(rep("P.val:",3), round(c(P1,P2,P3), 4)), pos=c(3,1,1), offset=1.3);


kableExtra::kbl(rbind(
" Box plot is showing distrubution of sample distances to the center of the cluster. Clusters members are defined by NCI_METRIC label only. ",

" Z: The Z-score number calculated for the geometrical distance of the tested sample (red dot) to the center of the cluster (middle of class labell). The center of the clusters calculated as a mean of x and y coordinates for all samples with same label. This Z score for the sample can be interpreted as a distance to the cluster center measured in standard deviations. In other words Z-score equal 1 means this sample is placed in about of one standard deviation from the cluster center.",

" P.val: is converted the Z.score under assumption those distances to cluster center are normaly distributed. Conversion done with 'pnorm' function. The smaller p.value the less chances the sample belongs to the cluster. Mislabeling of the cluster members can severily affect acuracy of p-value calulation.")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped","hover", "condensed"), font_size = 8)

DSTXT

```
